### PR TITLE
feat: Introduce two-level caching

### PR DESCRIPTION
**Pull request by Xuanwo**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

The cache policy still needs improvement, but I believe it's a good start.

## Setup

Aliyun 32C32G.

On a not very fast disk.

```
root@iZt4n8iblf9qk34w2qlvteZ:~/databend# fio -filename=./test_file -direct=1 -iodepth 1 -thread -rw=randrw -rwmixread=70 -ioengine=psync -bs=16k -size=2G -numjobs=10 -runtim$
=60 -group_reporting -name=test_r_w
test_r_w: (g=0): rw=randrw, bs=(R) 16.0KiB-16.0KiB, (W) 16.0KiB-16.0KiB, (T) 16.0KiB-16.0KiB, ioengine=psync, iodepth=1
...
fio-3.28
Starting 10 threads
Jobs: 10 (f=10): [m(10)][100.0%][r=39.6MiB/s,w=16.6MiB/s][r=2533,w=1065 IOPS][eta 00m:00s]
test_r_w: (groupid=0, jobs=10): err= 0: pid=281535: Wed Dec  7 15:46:39 2022
  read: IOPS=2560, BW=40.0MiB/s (41.9MB/s)(2400MiB/60006msec)
    clat (usec): min=140, max=20488, avg=2763.74, stdev=3726.97
     lat (usec): min=140, max=20488, avg=2763.80, stdev=3726.97
    clat percentiles (usec):
     |  1.00th=[  347],  5.00th=[  383], 10.00th=[  404], 20.00th=[  437],
     | 30.00th=[  461], 40.00th=[  482], 50.00th=[  510], 60.00th=[  553],
     | 70.00th=[  644], 80.00th=[ 8717], 90.00th=[ 8848], 95.00th=[ 8979],
     | 99.00th=[ 9110], 99.50th=[ 9241], 99.90th=[10290], 99.95th=[11469],
     | 99.99th=[19006]
   bw (  KiB/s): min=35456, max=123904, per=100.00%, avg=40971.56, stdev=785.21, samples=1190
   iops        : min= 2216, max= 7744, avg=2560.72, stdev=49.08, samples=1190
  write: IOPS=1099, BW=17.2MiB/s (18.0MB/s)(1031MiB/60006msec); 0 zone resets
    clat (usec): min=294, max=19663, avg=2656.44, stdev=3687.16
     lat (usec): min=294, max=19664, avg=2656.86, stdev=3687.16
    clat percentiles (usec):
     |  1.00th=[  334],  5.00th=[  351], 10.00th=[  359], 20.00th=[  367],
     | 30.00th=[  379], 40.00th=[  392], 50.00th=[  408], 60.00th=[  441],
     | 70.00th=[  545], 80.00th=[ 8586], 90.00th=[ 8717], 95.00th=[ 8848],
     | 99.00th=[ 8979], 99.50th=[ 9110], 99.90th=[ 9503], 99.95th=[ 9765],
     | 99.99th=[13304]
   bw (  KiB/s): min=12128, max=54976, per=100.00%, avg=17614.79, stdev=402.58, samples=1190
   iops        : min=  758, max= 3436, avg=1100.92, stdev=25.16, samples=1190
  lat (usec)   : 250=0.07%, 500=52.68%, 750=19.28%, 1000=0.51%
  lat (msec)   : 2=0.16%, 4=0.07%, 10=27.12%, 20=0.09%, 50=0.01%
  cpu          : usr=0.07%, sys=0.29%, ctx=219656, majf=0, minf=0
  IO depths    : 1=100.0%, 2=0.0%, 4=0.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=153617,65989,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=1

Run status group 0 (all jobs):
   READ: bw=40.0MiB/s (41.9MB/s), 40.0MiB/s-40.0MiB/s (41.9MB/s-41.9MB/s), io=2400MiB (2517MB), run=60006-60006msec
  WRITE: bw=17.2MiB/s (18.0MB/s), 17.2MiB/s-17.2MiB/s (18.0MB/s-18.0MB/s), io=1031MiB (1081MB), run=60006-60006msec

Disk stats (read/write):
  vda: ios=153328/66065, merge=0/49, ticks=422826/175754, in_queue=598583, util=99.80%
```

### Only Enable Fuse Meta Cache(Default)

Enable the Fuse Cache:

```
table_cache_enabled = true
```

Disable the block cache:

```
export STORAGE_CACHE_TYPE=none
```

| query | 1st(Cold)  | 2nd(Hot)   | 3rd(Hot)   |
| ----- | ---------- | ---------- | ---------- |
| q1    | 0.036475s  | 0.034330s  | 0.033860s  |
| q2    | 0.405285s  | 0.425448s  | 0.405614s  |
| q3    | 0.482817s  | 0.505521s  | 0.641434s  |
| q4    | 0.499226s  | 0.565254s  | 0.499185s  |
| q5    | 4.591048s  | 4.641461s  | 4.627463s  |
| q6    | 2.880130s  | 2.768725s  | 2.862527s  |
| q7    | 0.448876s  | 0.403117s  | 0.395856s  |
| q8    | 0.444990s  | 0.415852s  | 0.416349s  |
| q9    | 1.309182s  | 1.233529s  | 1.285408s  |
| q10   | 1.552085s  | 1.507529s  | 1.408535s  |
| q11   | 1.235884s  | 1.146333s  | 1.110571s  |
| q12   | 1.171811s  | 1.448887s  | 1.106544s  |
| q13   | 1.181518s  | 1.121305s  | 1.093806s  |
| q14   | 2.014658s  | 1.989183s  | 1.948704s  |
| q15   | 1.491372s  | 1.468302s  | 1.447846s  |
| q16   | 1.304342s  | 1.374811s  | 1.346403s  |
| q17   | 3.264561s  | 3.240071s  | 3.263775s  |
| q18   | 2.695813s  | 2.634235s  | 2.584193s  |
| q19   | 6.670163s  | 6.639486s  | 6.507481s  |
| q20   | 0.155114s  | 0.121644s  | 0.088172s  |
| q21   | 3.546669s  | 3.656849s  | 3.552669s  |
| q22   | 3.860203s  | 3.845587s  | 4.168806s  |
| q23   | 7.651057s  | 7.943695s  | 7.989370s  |
| q24   | 26.020538s | 22.834948s | 23.258524s |
| q25   | 1.370762s  | 1.356892s  | 1.378027s  |
| q26   | 0.729762s  | 0.815815s  | 0.678436s  |
| q27   | 1.396373s  | 1.543097s  | 1.387461s  |
| q28   | 3.548696s  | 3.508477s  | 3.612818s  |
| q29   | 6.181733s  | 6.261909s  | 6.071892s  |
| q30   | 1.299608s  | 1.234531s  | 1.228019s  |
| q31   | 1.713323s  | 1.645558s  | 1.637541s  |
| q32   | 2.416340s  | 2.529084s  | 2.402624s  |
| q33   | 13.003390s | 11.816549s | 11.737593s |
| q34   | 5.694307s  | 5.897234s  | 5.655931s  |
| q35   | 9.171278s  | 9.242920s  | 9.373954s  |
| q36   | 2.034718s  | 2.029365s  | 2.003341s  |
| q37   | 0.985171s  | 0.732181s  | 0.831386s  |
| q38   | 0.890445s  | 0.419350s  | 0.392929s  |
| q39   | 0.535444s  | 0.528808s  | 0.431822s  |
| q40   | 1.729394s  | 2.310463s  | 2.380718s  |
| q41   | 0.420815s  | 0.466013s  | 0.277656s  |
| q42   | 0.523699s  | 0.373747s  | 0.322487s  |
| q43   | 0.231658s  | 0.349807s  | 0.361606s  |

### With two level caching

Enable the Fuse Cache:

```
table_cache_enabled = true
```

Enable the block cache (with internal in-memory 1GiB)

```
export STORAGE_CACHE_TYPE=fs
export STORAGE_CACHE_FS_DATA_PATH=/root/databend/cache_data
```

| query  | 1st(Cold) | 2nd(Hot) | 3rd(Hot) |                                                               
| ------- | ------- | ------- | ------- |
| q1 | 0.109227s | 0.023928s | 0.022782s|
| q2 | 3.605249s | 0.549543s | 0.529290s|
| q3 | 0.982085s | 0.568819s | 0.585965s|
| q4 | 0.957540s | 0.729147s | 0.495203s|
| q5 | 4.519727s | 4.570618s | 4.453045s|
| q6 | 3.113226s | 3.074147s | 2.816269s|
| q7 | 0.659873s | 0.638950s | 0.446453s|
| q8 | 0.212356s | 0.249438s | 0.293220s|
| q9 | 1.234002s | 1.313146s | 1.218157s|
| q10 | 1.372646s | 1.385159s | 1.328493s|
| q11 | 1.167348s | 1.237570s | 0.895790s|
| q12 | 1.043435s | 0.959279s | 0.872513s|
| q13 | 1.188781s | 1.063294s | 0.922987s|
| q14 | 1.576568s | 1.681351s | 1.657352s|
| q15 | 1.288992s | 1.394368s | 1.254417s|
| q16 | 1.222763s | 1.209455s | 1.221277s|
| q17 | 3.128190s | 3.047233s | 3.012679s|
| q18 | 2.467448s | 2.436205s | 2.486016s|
| q19 | 6.396289s | 6.846603s | 6.404149s|
| q20 | 1.563110s | 0.061192s | 0.057500s|
| q21 | 4.956002s | 5.904866s | 8.215292s|
| q22 | 4.719741s | 4.533424s | 4.585998s|
| q23 | 9.580432s | 11.910211s | 9.107909s|
| q24 | 22.469498s | 19.955859s | 19.799339s|
| q25 | 0.659339s | 0.589525s | 0.586172s|
| q26 | 0.381847s | 0.320696s | 0.346925s|
| q27 | 0.591120s | 0.592103s | 0.587410s|
| q28 | 3.156763s | 3.125415s | 3.167056s|
| q29 | 6.071926s | 7.279704s | 6.875829s|
| q30 | 1.204553s | 1.216558s | 1.220133s|
| q31 | 1.459237s | 1.367377s | 1.412679s|
| q32 | 2.226188s | 2.141515s | 2.583990s|
| q33 | 13.927794s | 12.531810s | 12.368160s|
| q34 | 15.852458s | 5.269190s | 5.295076s|
| q35 | 8.862489s | 8.932716s | 9.153747s|
| q36 | 2.030967s | 1.913406s | 1.894756s|
| q37 | 0.875297s | 0.644295s | 0.639375s|
| q38 | 0.801089s | 0.433443s | 0.432999s|
| q39 | 0.447179s | 0.288095s | 0.331832s|
| q40 | 1.743530s | 1.752901s | 1.735988s|
| q41 | 0.338641s | 0.292702s | 0.266163s|
| q42 | 0.307983s | 0.226774s | 0.185038s|
| q43 | 0.155204s | 0.117806s | 0.124273s|

### 1st run (no cache vs fs cache)

image

### 2nd run (no cache vs fs cache)

image

### 3nd run (no cache vs fs cache)

image

### fs cache under 1/2/3

image

### Metrics

image

### CODE REVIEW

File: src/common/storage/src/cache.rs [hunk 1]
```diff
@@ -59,46 +117,87 @@ impl CachePolicy for FuseCachePolicy {
...
+            if size < 4 * 1024 * 1024 {
+                let mut bs = Vec::with_capacity(size as usize);
+                r.read_to_end(&mut bs).await.map_err(|err| {
+                    Error::new(ErrorKind::Unexpected, "read from underlying storage")
+                        .set_source(err)
+                })?;
+                let bs = Bytes::from(bs);
+
+                if enable_async {
+                    let moved_bs = bs.clone();
```
Comment (BohuTANG) [inline src/common/storage/src/cache.rs:163]: How about the memory usage if enable_async enabled?

File: src/common/storage/src/cache.rs [hunk 2]
```diff
@@ -29,27 +36,78 @@ use opendal::OpRead;
...
 
     fn cache_path(&self, path: &str, args: &OpRead) -> String {
-        format!("{path}.cache-{}", args.range().to_header())
+        format!("{path}.cache-{}", args.range())
     }
 }
 
 #[async_trait]
-impl CachePolicy for FuseCachePolicy {
+impl CachePolicy for RangeCachePolicy {
```
Comment (BohuTANG) [inline src/common/storage/src/cache.rs:110]: Move this to a single file like `range_cache.rs`?  We can add more about how the range cache policies work to this file.

File: src/common/storage/src/cache.rs (hunk 1 above)
Comment (Xuanwo) [reply src/common/storage/src/cache.rs:163]: No big different as we only have `concurrent` running async tasks for cache.

File: src/common/storage/src/cache.rs (hunk 2 above)
Comment (BohuTANG) [reply src/common/storage/src/cache.rs:110]: Oh, the file name `cache.rs` is ok.
Add more about how the cache work, like when the lay1 is full and evicted to the lay2.

**Review by PsiACE — APPROVED** [review]

**Comment by Xuanwo** [comment]
image

I'm working to avoid too much cache missing in moka.

**Comment by Xuanwo** [comment]
Data has been updated.

**Comment by Xuanwo** [comment]
q21,q23,q33 not looks good, I'm investigating.

**Comment by Xuanwo** [comment]
Our bench has been affected by cluster mode, re-bench again.

**Comment by BohuTANG** [comment]
I have tested this patch again; it looks like the cache policy is strange, it can read any data from my local disk  cache.

## Query config
```
table_cache_enabled = true
export STORAGE_CACHE_TYPE=fs
export STORAGE_CACHE_FS_DATA_PATH=/root/databend/cache_data
```

## Query
```
SELECT * FROM hits WHERE URL LIKE '%google%' ORDER BY EventTime LIMIT 10;
```

## First time(Load data from remote s3):
image

Cache data size:
```
du -h cache_data/*
0       cache_data/health_check
```

## Second time(**Load data from remote s3**):
image

Cache data size:
```
du -h cache_data/*
217M    cache_data/355227/355233/_b
217M    cache_data/355227/355233
217M    cache_data/355227
0       cache_data/health_check
```

## Third time(**Still load data from remote s3**):
image


Cache data size:
```
du -h cache_data/*
467M    cache_data/355227/355233/_b
467M    cache_data/355227/355233
467M    cache_data/355227
0       cache_data/health_check
```

**Comment by BohuTANG** [comment]
This way may be efficient:
Write:
The level-1 memory cache is the write buffer. If it flushes data to disk(level 2) when evicting, level 1 only keeps the key, and removes the value.
Read:
Try to get the key from the fast level 0. If it was cached, read the data from level 2(For example, from fs disk).

**Comment by Xuanwo** [comment]
> Write:
> The level-1 memory cache is the write buffer. If it flushes data to disk(level 2) when evicting, level 1 only keeps the key, and removes the value.
> Read:
> Try to get the key from the fast level 1. If it was cached, read the data from level 2(For example, from fs disk).

Let me try this.

**Pull request closed without merging**